tests/integration/conformance/conformance_test.go

```diff
@@ -151,3 +171,192 @@ func validateMCPState(ctx test.Failer, gal galley.Instance, ns namespace.Instanc
+		for _, domain := range domains {
+			// TODO(qfel): Figure out how to get rid of the loop.
+			if err := validator.Exists(q, domain).Check(); err != nil {
+				return false, err
+			}
+		}
+		return true, nil
+	}
+}
+
+func virtualServiceHosts(ctx test.Failer, resources string) []string {
+	// TODO(qfel): We should parse input early and work on structured data here. Or at least if some
+	// other code path needs it.
+	var inputs []map[interface{}]interface{}
+	for _, inputYAML := range strings.Split(resources, "\n---\n") {
+		var input map[interface{}]interface{}
+		if err := yaml.Unmarshal([]byte(inputYAML), &input); err != nil {
+			ctx.Fatalf("yaml.Unmarshal: %v", err)
+		}
+		inputs = append(inputs, input)
+	}
+
+	var vHosts []string
+	for _, res := range inputs {
+		if res["apiVersion"] != "networking.istio.io/v1alpha3" || res["kind"] != "VirtualService" {
+			continue
+		}
+		spec := res["spec"].(map[interface{}]interface{})
+		hosts := spec["hosts"].([]interface{})
+		for _, h := range hosts {
+			vHosts = append(vHosts, h.(string))
+		}
+	}
+	return vHosts
+}
+
+func validateTraffic(t framework.TestContext, pil pilot.Instance, gal galley.Instance, ns namespace.Instance, stage *conformance.Stage) {
+	echos := make([]echo.Instance, len(stage.Traffic.Services))
+	b := echoboot.NewBuilderOrFail(t, t)
+	for i, svc := range stage.Traffic.Services {
+		var ports []echo.Port
+		for _, p := range svc.Ports {
+			ports = append(ports, echo.Port{
+				Name:        p.Name,
+				Protocol:    model.Protocol(p.Protocol),
+				ServicePort: int(p.ServicePort),
+			})
+		}
+		b = b.With(&echos[i], echo.Config{
+			Galley:    gal,
+			Pilot:     pil,
+			Service:   svc.Name,
+			Namespace: ns,
+			Ports:     ports,
+		})
+	}
+	if err := b.Build(); err != nil {
+		t.Fatal(err)
+	}
+
+	services := make(map[string]echo.Instance)
+	for i, svc := range echos {
+		services[stage.Traffic.Services[i].Name] = svc
+		svc.WaitUntilCallableOrFail(t, echos...)
+	}
+
+	ready := make(map[string]bool)
+	vHosts := virtualServiceHosts(t, stage.Input)
+	for _, call := range stage.Traffic.Calls {
+		call.URL = tmpl.EvaluateOrFail(t, call.URL, constraint.Params{
+			Namespace: ns.Name(),
+		})
+		t.NewSubTest(call.URL).Run(func(t framework.TestContext) {
+			hostname, err := wildcardToRegexp(call.Response.Callee)
+			if err != nil {
+				t.Fatalf("Internal error: regexp.Compile: %v", err)
+			}
+			caller := services[call.Caller]
+			if !ready[call.Caller] && false {
+				t.Logf("Waiting for sidecar(s) for %s to contain domains: %s", call.Caller, strings.Join(vHosts, ", "))
+				for _, w := range caller.WorkloadsOrFail(t) {
+					w.Sidecar().WaitForConfigOrFail(t, domainAcceptFunc(vHosts))
+				}
+				ready[call.Caller] = true
+			}
+			u, err := url.Parse(call.URL)
+			if err != nil {
+				t.Fatalf("Parse call URL: %v", err)
+			}
+			workload := caller.WorkloadsOrFail(t)[0]
+
+			validateWithRedo(t, func(ctx context.Context) bool {
```

Comment:
There are utility methods for doing retry in the test/retry package (IIRC). Could you use them?

Reply:
tests/util/retry.go, pkg/test/util/retry/retry.go or doesn't matter?

Reply:
> pkg/test/util/retry

It's not context-aware and does not implement the entire try-multiple-times strategy, so I'd say there's not much point in trying to use it. If we did use it in its current form, the could could get even more complicated.